daemons/schedulerd/pacemaker-schedulerd.c

```diff
@@ -369,21 +369,24 @@ main(int argc, char **argv)
     mainloop = g_main_loop_new(NULL, FALSE);
     crm_notice("Pacemaker scheduler successfully started and accepting connections");
     g_main_loop_run(mainloop);
-
-    pe_free_working_set(sched_data_set);
-    crm_info("Exiting %s", crm_system_name);
-
-    pcmk__unregister_formats();
-    out->finish(out, CRM_EX_OK, true, NULL);
-    pcmk__output_free(out);
-
-    crm_exit(CRM_EX_OK);
+    pengine_shutdown(0);
 }
 
 void
 pengine_shutdown(int nsig)
 {
     mainloop_del_ipc_server(ipcs);
+    ipcs = NULL;
+
     pe_free_working_set(sched_data_set);
+    sched_data_set = NULL;
+
+    if (out != NULL) {
+        pcmk__unregister_formats();
```

Comment:
You can call pcmk__unregister_formats even if out is NULL.  That function already does a check before it tries to do anything.

Reply:
Good point, and theoretically pcmk__output_new() could fail and we'd still want to unregister the formats. Done.